### PR TITLE
M0078DDG-331 Add coding rules tutorial

### PR DESCRIPTION
**Pull request by apoudret**

### CODE REVIEW

**Comment by frivieremicroej** [comment]
Public link: https://docs.microej.com/en/feature-m0078ddg-331_coding_rules/

File: Tutorials/tutorialJavaCodingRules.rst [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+- ``equals(Object)`` and ``hashCode()`` must be overridden in
+  pairs. See :ref:`equals-hashcode`.
+- Do not assign fields in field declaration but in the constructor.
+- Do not use non-final method inside the constructor.
+- Do not overburden the constructor with logic.
+- Prefer a default constructor (without parameters) in order to use
+  reflection (``Class.newInstance()``).
+- Do not directly store an array given by parameter.
+- Do not directly return an internal array.
+- Save object reference from a field to a local before using it::
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:162]: good, Improve the text (TODO friviere)

caching the object in a local variable is always a good practice => race conditions, performance improvements (loop invariants...), footprint reduction, naming (documentation)

File: Tutorials/tutorialJavaCodingRules.rst [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+--------------------------------------
+
+- Extract constants instead of using magic numbers.
+- Use parenthesis for complex operation series; it simplifies the understanding 
+  of operator priorities.
+- Write short lines. This can be achieved by extracting variables (for
+  example: (``(a == null || b == null)`` becomes
+  ``(aIsNull || bIsNull)``).
+- Use component-oriented architecture to separate concerns (Use
+  ej.library.runtime#component).
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:179]: TODO reformulate  service ? add a link as usual

File: Tutorials/tutorialJavaCodingRules.rst [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+- Use a limited number of parameters in methods (or perhaps a new type
+  is needed).
+- Create small methods with little complexity. When a method gets too
+  complex, it should be split.
+- Use ``+`` operator for single-line string concatenation. Use a
+  StringBuilder otherwise.
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:185]: javadoc link

File: Tutorials/tutorialJavaCodingRules.rst [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+Performance considerations
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:187]: to be moved elsewhere (performance tuto ?)

File: Tutorials/tutorialJavaCodingRules.rst [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+Basic optimizations
+-------------------
+
+- Avoid initializing fields to ``0`` or ``null``. A ``//VM_DONE`` tag
+  can be written to understand the optimization.
+- The switch/case blocks are generated in two ways depending on the
+  cases. Prefer consecutive cases for faster execution (and slightly
+  smaller bytecode).
+- Avoid using synchronized types (Vector, Hashtable, StringBuffer,
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:209]: OK TODO friviere reformulate

Comment (frivieremicroej) [reply Tutorials/tutorialJavaCodingRules.rst:209]: apoudret & seon this is a very good and helpful document
let's have a call to discuss about my comments

File: Tutorials/tutorialJavaCodingRules.rst [hunk 6]
```diff
@@ -0,0 +1,332 @@
+.. _javacodingrules:
+
+Improve the Quality of the Code of an Application
+=================================================
+
+This tutorial explains how to analyze the quality of the code of an application and how to improve this quality in order to simplify the maintenance of the code.
```
Comment (frivieremicroej) [inline Tutorials/tutorialJavaCodingRules.rst:6]: TODO friviere intro which summarize amounts of tips that have been experienced by MicroEJ to both code quality & performance for embedded systems.

File: Tutorials/tutorialJavaCodingRules.rst (hunk 2 above)
Comment (apoudret) [reply Tutorials/tutorialJavaCodingRules.rst:179]: removed reference to the library, the rule is now generic. the footprint guide now suggests using `Constants.getClass()`

File: Tutorials/tutorialJavaCodingRules.rst [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+  - Instance variables. First, the public class variables, then the
+    protected, then package level (no access modifier), and then the
+    private.
+  - Constructors
+  - Methods
+
+Bad Practices
+-------------
+
+- Avoid redefining ``toString()`` method for debugging purposes. The method
```
[the remark this answers is not in the document]
Comment (apoudret) [reply Tutorials/tutorialJavaCodingRules.rst:145]: moved to footprint tuto

**Pull request merged**